### PR TITLE
feat: bytesPerSecond function generated by request, allow async bytePerSecond functions

### LINKED ISSUE #7: Possibility to retrieve request object in bytesPerSecond function
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

It’d be great if we could access the request object in the bytesPerSecond function so we can set the speed based on outer arguments.

### Motivation

A possible use case could be throttling speeds for different user subscription tiers.

### Example

Let’s say I have a `req.user.plan` field and want to throttle users depending on which plan they have, i currently see no way to implement this behavior. If the `req` object could be passed that’d be awesome.

### PR DESCRIPTION
**Pull request by Uzlopak**
Resolves #7 

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by Fdawgs — APPROVED** [review]

File: index.js [hunk 1]
```diff
@@ -36,32 +36,59 @@ async function addRouteThrottleHook (fastify, routeOptions, throttleOptions) {
...
 
-  return function onSendHandler (request, reply, payload, done) {
-    if (throttleOpts.streamPayloads && payload instanceof Stream) {
-      done(null, pipeline(
-        payload,
-        new ThrottleStream({ bytesPerSecond }),
-        err => { fastify.log.error(err) }
-      ))
-      return
+  if (typeof bytesPerSecond === 'number') {
```
Comment (Eomm) [inline index.js:39]: I would keep the async style only - much more easier to read. 
Or the performance would be awful at that point? 😄

Comment (Uzlopak) [reply index.js:39]: You mean check if it is an async function and then only await?

File: README.md [hunk 2]
```diff
@@ -120,7 +122,7 @@ the `bytesPerSecond` like this:
   fastify.get('/', {
     config: {
       throttle: {
-        bytesPerSecond: function (elapsedTime, bytes) {
+        bytesPerSecond: (request) => function (elapsedTime, bytes) {
```
[the remark this answers is not in the document]
Comment (Uzlopak) [reply README.md:125]: hmm. it is clear for me. :D

but serious: Can you provide a suggestion please?

[the remark this answers is not in the document]
Comment (Eomm) [reply README.md:125]: nood style:

```
bytesPerSecond: function factory(request) {
  // this function runs for every request
  const client = request.headers['customer-id']
  
  return function (elapsedTime, bytes) {
     return CONFIG[client] * 2 // return a number of xyz
  }
}
```

File: index.js (hunk 1 above)
Comment (Eomm) [reply index.js:39]: use always `async onSendHandler` and await `bytesPerSecond` if necessary

Comment (Uzlopak) [reply index.js:39]: then I have to detect if it is an async function. Well I will probably restrict it to only async function but not functions returning Promises.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by gurgunday — APPROVED** [review]

**Comment by gurgunday** [comment]
FYI, this _core_ plugin isn't under the team fastify/plugins

@Uzlopak

**Comment by Uzlopak** [comment]
Good catch. Added fastify/plugins team with maintainer permissions.

**Comment by Uzlopak** [comment]
Added an option async to allow sync functions returning promises.

@Fdawgs 
Please have a look regarding the readme.

@Eomm 
@gurgunday 
Please check if this is better

**Review by Fdawgs — APPROVED** [review]

**Comment by Uzlopak** [comment]
Well. I am the creator and i am confident in my changes (would worry if otherwise, lol).  So i would merge it. :p

But i have no permissions to release it on npm. ....

@gurgunday 

Can you release?

**Review by gurgunday — APPROVED** [review]

**Comment by Uzlopak** [comment]
@gurgunday 

Thanks ;)

**Comment by Uzlopak** [comment]
@gurgunday 
Please release it as 2.0.0

**Pull request merged**